cmd/node.go

```diff
@@ -1,67 +1,131 @@
-// down each component if a SIGINT is received. Until a SIGINT is received, Run will block.
-// Since, Run is a blocking call it should only be used when running a node as it's own independent process.
+// Run starts all the node's components, then blocks until a SIGINT or SIGTERM is received, at
+// which point it gracefully shuts down.
+// Any unhandled irrecoverable errors thrown in child components will propagate up to here and
+// result in a fatal error.
 func (node *FlowNodeImp) Run() {
-	// initialize signal catcher
-	signal.Notify(node.sig, os.Interrupt, syscall.SIGTERM)
-
-	select {
-	case <-node.Ready():
-		node.Logger.Info().Msgf("%s node startup complete", node.BaseConfig.NodeRole)
-	case <-node.sig:
-		node.Logger.Warn().Msg("node startup aborted")
-		os.Exit(1)
+	// Block until node is shutting down
+	err := node.run()
+
+	// Any error received is considered fatal.
+	if err != nil {
+		node.fatalHandler(err, node.logger)
+		return
+	}
+
+	// Run post shutdown cleanup logic
+	err = node.postShutdown()
+
+	// Since this occurs after all components have stopped, it is not considered fatal
+	if err != nil {
+		node.logger.Error().Err(err).Msg("error encountered during cleanup")
 	}
 
-	// block till a SIGINT is received
-	<-node.sig
+	node.logger.Info().Msgf("%s node shutdown complete", node.BaseConfig.NodeRole)
+}
+
+// run starts the node and blocks until a SIGINT/SIGTERM is received or an error is encountered.
+// It returns:
+//   - nil if a termination signal is received, and all components have been gracefully stopped.
+//   - error if a irrecoverable error is received
+func (node *FlowNodeImp) run() error {
+	// Cancelling this context notifies all child components that it's time to shutdown
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Components will pass unhandled irrecoverable errors to this channel via signalerCtx (or a
+	// child context). Any errors received on this channel should halt the node.
+	signalerCtx, errChan := irrecoverable.WithSignaler(ctx)
 
-	node.Logger.Info().Msgf("%s node shutting down", node.BaseConfig.NodeRole)
+	// This context will be marked done when SIGINT/SIGTERM is received.
+	sigCtx, stop := signal.NotifyContext(ctx, os.Interrupt, syscall.SIGTERM)
 
-	select {
-	case <-node.Done():
-		node.Logger.Info().Msgf("%s node shutdown complete", node.BaseConfig.NodeRole)
-	case <-node.sig:
-		node.Logger.Warn().Msg("node shutdown aborted")
-		os.Exit(1)
+	// 1: Start up
+	// Start all the components
+	node.Start(signalerCtx)
+
+	// Log when all components have been started
+	go func() {
+		select {
+		case <-node.Ready():
+			node.logger.Info().Msgf("%s node startup complete", node.BaseConfig.NodeRole)
+		case <-ctx.Done():
+		}
+	}()
+
+	// 2: Run the node
+	// Block here until either a signal or irrecoverable error is received.
+	err := util.WaitError(sigCtx, errChan)
+
+	// Stop relaying signals. Subsequent signals will be handled by the OS and will abort the
+	// process.
```

Comment:
It means if we hit an irrecoverable error, we are going to stop all components. And during stopping them, if we receive a SigTerm signal again, it will handled by the OS and exit right away?

Reply:
yea

Reply:
nice, this is a good way to handle it.